lib/Dialect/Seq/Transforms/LowerSeqHLMem.cpp

```diff
@@ -0,0 +1,156 @@
+    llvm::SmallVector<seq::ReadPortOp> readOps;
+    llvm::SmallVector<seq::WritePortOp> writeOps;
+    for (auto *user : mem.getHandle().getUsers()) {
+      auto res = llvm::TypeSwitch<Operation *, LogicalResult>(user)
+                     .Case([&](seq::ReadPortOp op) {
+                       readOps.push_back(op);
+                       return success();
+                     })
+                     .Case([&](seq::WritePortOp op) {
+                       writeOps.push_back(op);
+                       return success();
+                     })
+                     .Default([&](Operation *op) { return failure(); });
+      if (failed(res))
+        return rewriter.notifyMatchFailure(user, "unsupported port type");
+    }
+
+    auto clk = mem.getClk();
+    auto rst = mem.getRst();
+    auto memName = mem.getName();
+
+    // Create the SV memory.
+    hw::UnpackedArrayType memArrType =
+        hw::UnpackedArrayType::get(memType.getElementType(), size);
+    auto svMem =
+        rewriter.create<sv::RegOp>(mem.getLoc(), memArrType, mem.getNameAttr())
+            .getResult();
+
+    // Create write ports by gathering up the write port inputs and
+    // materializing the writes inside a single always ff block.
+    struct WriteTuple {
+      Location loc;
+      Value addr;
+      Value data;
+      Value en;
+    };
+    llvm::SmallVector<WriteTuple> writeTuples;
+    for (auto writeOp : writeOps) {
+      if (writeOp.getLatency() != 1)
+        return rewriter.notifyMatchFailure(
+            writeOp, "only supports write ports with latency == 1");
+      auto addr = writeOp.getAddresses()[0];
+      auto data = writeOp.getInData();
+      auto en = writeOp.getWrEn();
+      writeTuples.push_back({writeOp.getLoc(), addr, data, en});
+      rewriter.eraseOp(writeOp);
+    }
+
+    rewriter.create<sv::AlwaysFFOp>(
+        mem.getLoc(), sv::EventControl::AtPosEdge, clk, ResetType::SyncReset,
+        sv::EventControl::AtPosEdge, rst, [&] {
+          for (auto [loc, address, data, en] : writeTuples) {
+            Value a = address, d = data; // So the lambda can capture.
+            Location l = loc;
+            // Perform write upon write enable being high.
+            rewriter.create<sv::IfOp>(loc, en, [&] {
+              Value memLoc =
+                  rewriter.create<sv::ArrayIndexInOutOp>(l, svMem, a);
+              rewriter.create<sv::PAssignOp>(l, memLoc, d);
+            });
+          }
+        });
+
+    // Create read ports.
+    for (auto [ri, readOp] : llvm::enumerate(readOps)) {
+      rewriter.setInsertionPointAfter(readOp);
+      auto loc = readOp.getLoc();
+
+      // Create a combinational read.
+      Value memLoc = rewriter.create<sv::ArrayIndexInOutOp>(
+          loc, svMem, readOp.getAddresses()[0]);
+      Value readData = rewriter.create<sv::ReadInOutOp>(loc, memLoc);
+
+      // Materialize any delays on the read port.
+      for (int i = 0, e = readOp.getLatency(); i < e; ++i)
```

Comment:
If the latency > 1, there should be at least one pipeline stage on the address side. Maybe half on one side? Actually, it may make sense to have separate latencies to be able to more precisely resolve RAW hazards.

Reply:
> able to more precisely resolve RAW hazards

IMO, this op should either provide explicit guarantees about how RUW is handled, or none at all (undefined). Resolving conflicts "more precisely" is still not deterministic.

> If the latency > 1, there should be at least one pipeline stage on the address side.

My view of correctness here is that a read is always performed semantically in the cycle which the access occurred. I think it's either that, or provide no guarantees at all (reads can be performed at any point within the provided latency - would this your "at least one pipeline stage"?).

> Maybe half on one side?

as in, to support single-cycle RAW conflict handling, i assume. I'm not sure whether this should be the default. Things which need this (e.g. register files) afaik tend to be specially designed, and must be so, if you don't want to kill your fmax by now having only half a cycle of slack available on the memory read signal.

Reply:
Will merge this just to get things moving downstream of this commit for today, but we can of course revisit this part of the lowering.